File: internal/authz/judgment_policy.go
// Fiscal convention: monetary values in the Drenyra ecosystem are int64 cents;
// no float is ever used for money. This module is the versioned, PURE judgment
// authorization policy (v0.4.0 Step 2 — adjudicable conflicts). Like the
// approval policy, it works from the PRE-VERIFIED principal only — no database,
// clock, token, or identity-provider access. The judgment row records the
// provenance; this policy authorizes confirmation/rejection.
//
// Frozen judgment-policy/v0.4.0 matrix (design §6):
//   - Minimum role is senior_accountant; controller dominates it through the
//     accounting ladder; tax roles do NOT (a controller never implies
//     tax_reviewer/authorized_tax_professional, and vice versa).
//   - Minimum assurance is standard.
//   - The judgment must be tenant-scoped to the principal's tenant and its
//     company must be inside the membership's companyScopes (a non-empty
//     company is required).
//   - Check order (first frozen code wins, for reproducibility): tenant →
//     company scope → membership active → role → assurance.
package authz

import (
	"github.com/arkelythex/drenyra-engram/internal/auth"
	"github.com/arkelythex/drenyra-engram/internal/core"
)

// JudgmentPolicyVersion is the frozen version of the judgment policy. It is
// stamped on every decision and persisted with confirm/reject events.
const JudgmentPolicyVersion = "judgment-policy/v0.4.0"

// JudgmentAuthorizationDecision is the pure authorization outcome: allowed, the
// exact policy version, and the frozen reason code (AUTHORIZED when allowed).
type JudgmentAuthorizationDecision struct {
	Allowed       bool
	PolicyVersion string
	ReasonCode    string
}

// JudgmentAuthorizationPolicy is the pure authorization contract for
// confirming/rejecting a judgment. Implementors must not touch the database,
// the clock, tokens, or identity providers.
type JudgmentAuthorizationPolicy interface {
	Authorize(principal auth.VerifiedApprovalPrincipal, judgment core.AccountingJudgment) JudgmentAuthorizationDecision
}

// frozenJudgmentPolicy is the immutable judgment-policy/v0.4.0 policy.
type frozenJudgmentPolicy struct{}

// NewJudgmentPolicy returns the frozen judgment-policy/v0.4.0 policy.
func NewJudgmentPolicy() JudgmentAuthorizationPolicy { return frozenJudgmentPolicy{} }

// Authorize evaluates the pre-verified principal against the judgment's tenant,
// company, membership, role and assurance — in that exact order; the first
// frozen reason code wins so denials are reproducible.
func (frozenJudgmentPolicy) Authorize(p auth.VerifiedApprovalPrincipal, j core.AccountingJudgment) JudgmentAuthorizationDecision {
	// 1. Tenant scope: the principal's tenant must equal the judgment's tenant.
	if p.TenantID() != j.TenantID {
		return judgmentDenied(auth.CodeTenantScopeMismatch)
	}
	// 2. Company scope: the judgment's company must be inside the membership's
	//    company scopes. A judgment without a company ("" ) fails closed.
	if j.CompanyID == "" || !contains(p.CompanyScopes(), j.CompanyID) {
		return judgmentDenied(auth.CodeCompanyScopeDenied)
	}
	// 3. Membership active: a pre-verified principal always carries the
	//    membership id of the ACTIVE membership it was derived from; empty
	//    means no active membership.
	if p.MembershipID() == "" {
		return judgmentDenied(auth.CodeMembershipInactive)
	}
	// 4. Role: the principal must satisfy senior_accountant (minimum). Only the
	//    accounting ladder dominates (controller ≥ senior_accountant); tax
	//    roles sit outside the ladder and never authorize adjudication.
	if !satisfies(p.Roles(), auth.RoleSeniorAccountant) {
		return judgmentDenied(auth.CodeRoleNotAuthorized)
	}
	// 5. Assurance: minimum standard.
	if assuranceRank(p.AssuranceLevel()) < assuranceRank(auth.AssuranceStandard) {
		return judgmentDenied(auth.CodeAssuranceTooLow)
	}
	return JudgmentAuthorizationDecision{Allowed: true, PolicyVersion: JudgmentPolicyVersion, ReasonCode: ReasonAuthorized}
}

// judgmentDenied returns a denied decision carrying the frozen policy version.
func judgmentDenied(code string) JudgmentAuthorizationDecision {
	return JudgmentAuthorizationDecision{Allowed: false, PolicyVersion: JudgmentPolicyVersion, ReasonCode: code}
}

File: authz/judgment-policy.ts
/**
 * Fiscal convention: monetary values in the Drenyra ecosystem are BigInt cents;
 * no float is ever used for money.
 *
 * Versioned judgment authorization policy (v0.4.0 Step 2) — the exact
 * TypeScript mirror of `internal/authz/judgment_policy.go`. PURE: no database,
 * clock, token, or identity-provider access; it authorizes from the
 * pre-verified principal only.
 *
 * Frozen judgment-policy/v0.4.0 matrix and check order (first reason code wins,
 * for reproducibility): tenant → company scope → membership active → role →
 * assurance. Minimum role senior_accountant (controller dominates it through
 * the accounting ladder; tax roles never do); minimum assurance standard; the
 * judgment's company must be inside the membership's companyScopes.
 */

import type {
	AccountingJudgment,
	AccountingRole,
	ApprovalErrorCode,
	AssuranceLevel,
	VerifiedApprovalPrincipal,
} from "../core/types.js";
import { REASON_AUTHORIZED } from "./approval-policy.js";

/** The frozen policy version, stamped on every decision. */
export const JUDGMENT_POLICY_VERSION = "judgment-policy/v0.4.0";

/** Pure authorization outcome: allowed, the exact policy version, reason code. */
export interface JudgmentAuthorizationDecision {
	allowed: boolean;
	/** Exactly `judgment-policy/v0.4.0`. */
	policyVersion: string;
	/** Frozen reason code (or `AUTHORIZED` when allowed). */
	reasonCode: string;
}

/**
 * Accounting ladder rank: accountant < senior_accountant < controller. Tax roles
 * sit OUTSIDE the ladder at rank 0 and can never dominate it. (Mirror of the
 * ladder in approval-policy.ts; duplicated here so this module is
 * self-contained and the Step 1 file stays untouched.)
 */
const LADDER_RANK: Record<AccountingRole, number> = {
	accountant: 1,
	senior_accountant: 2,
	controller: 3,
	tax_reviewer: 0,
	authorized_tax_professional: 0,
};

const ASSURANCE_RANK: Record<AssuranceLevel, number> = {
	low: 0,
	standard: 1,
	strong: 2,
};

function judgmentDenied(
	reasonCode: ApprovalErrorCode,
): JudgmentAuthorizationDecision {
	return {
		allowed: false,
		policyVersion: JUDGMENT_POLICY_VERSION,
		reasonCode,
	};
}

/**
 * Role dominance exists ONLY within the accounting ladder; tax roles are
 * matched explicitly — a controller NEVER implies tax_reviewer or
 * authorized_tax_professional, and vice versa.
 */
function satisfies(
	roles: readonly AccountingRole[],
	required: AccountingRole,
): boolean {
	if (required === "tax_reviewer" || required === "authorized_tax_professional") {
		return roles.includes(required);
	}
	const rr = LADDER_RANK[required];
	return roles.some((role) => LADDER_RANK[role] >= rr);
}

/**
 * The pure judgment-policy/v0.4.0 decision for a pre-verified principal and a
 * judgment. Returns the FIRST frozen reason code in the fixed check order.
 */
export function authorizeJudgment(
	principal: VerifiedApprovalPrincipal,
	judgment: AccountingJudgment,
): JudgmentAuthorizationDecision {
	// 1. Tenant scope: the principal's tenant must equal the judgment's tenant.
	if (principal.tenantId !== judgment.tenantId) {
		return judgmentDenied("TENANT_SCOPE_MISMATCH");
	}
	// 2. Company scope: the judgment's company must be inside the membership's
	//    company scopes; a judgment without a company fails closed.
	if (
		judgment.companyId.length === 0 ||
		!principal.companyScopes.includes(judgment.companyId)
	) {
		return judgmentDenied("COMPANY_SCOPE_DENIED");
	}
	// 3. Membership active: a pre-verified principal always carries the
	//    membership id of the ACTIVE membership it was derived from; empty
	//    means no active membership.
	if (principal.membershipId === "") {
		return judgmentDenied("MEMBERSHIP_INACTIVE");
	}
	// 4. Role: the principal must satisfy senior_accountant (minimum). Only the
	//    accounting ladder dominates; tax roles never authorize adjudication.
	if (!satisfies(principal.roles, "senior_accountant")) {
		return judgmentDenied("ROLE_NOT_AUTHORIZED");
	}
	// 5. Assurance: minimum standard.
	if (ASSURANCE_RANK[principal.assuranceLevel] < ASSURANCE_RANK["standard"]) {
		return judgmentDenied("ASSURANCE_TOO_LOW");
	}
	return {
		allowed: true,
		policyVersion: JUDGMENT_POLICY_VERSION,
		reasonCode: REASON_AUTHORIZED,
	};
}
